New binding: panStamp wireless Arduino modules

GideonLeGrange opened the pull request:
A binding to support panStamp wireless Arduino modules. 

* I have successfully tested the binding on openhab 1.7.x and in compatibility mode on openhab 2.0 alpha2.
* I'm busy writing wiki documentation, but for the interim my testing documentation and configs are available here: https://github.com/AutoMates/openhab-panstamp-test

For more information about panStamp, please see: http://panstamp.com/
This library is based on panstamp-java, please see: https://github.com/GideonLeGrange/panstamp-java

GideonLeGrange commented:
Added Wiki page: https://github.com/openhab/openhab/wiki/panStamp-Binding

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,583 @@
+	}
+
+	@Override
+	protected void execute() {
```
If you don't use the polling of an AbstractActiveBinding, extend AbstractBinding instead.

GideonLeGrange commented:
@watou I have changed the inheritance as requested

watou commented:
Thank you.  I hope to offer more comments.  Is there a discussion thread?

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampGenericBindingProvider.java:
```diff
@@ -0,0 +1,155 @@
+		String unit = confMap.get("unit");
+		return new PanStampBindingConfig(itemName,
+				PanStampConversions.asInt("device address", confMap.get("address"), 1, 0xffff), 
+				parseManufacturerId(confMap.get("productCode")), 
```
^ Is this correct?

GideonLeGrange commented:
Yes, there is: https://community.openhab.org/t/new-binding-panstamp-wireless-arduino-modules/2731

watou commented on bundles/binding/org.openhab.binding.panstamp/pom.xml:
```diff
@@ -0,0 +1,35 @@
+	<properties>
+		<bundle.symbolicName>org.openhab.binding.panstamp</bundle.symbolicName>
+		<bundle.namespace>org.openhab.binding.panstamp</bundle.namespace>
+		<deb.name>openhab-addon-binding-PanStamp</deb.name>
```
use lower case: `openhab-addon-binding-panstamp`

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/PanStampBindingConfig.java:
```diff
@@ -0,0 +1,112 @@
+	private final String endpoint;
+	private final String unit;
+	
+}
```
add newline

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,566 @@
+
+import org.openhab.binding.panstamp.PanStampBindingConfig;
+import org.openhab.binding.panstamp.PanStampBindingProvider;
+import org.openhab.core.binding.AbstractActiveBinding;
```
remove import

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,566 @@
+
+			startup(PanStampBindingSettings.parseConfig(configuration));
+		} catch (ValueException e) {
+			logger.error(e.getMessage());
```
Possibly more descriptive error message?

GideonLeGrange replied:
The error messages generated by ```parseConfig()``` should usually be descriptive and tell the user which setting failed for what reason. I am changing this to: 

```java
catch (ValueException e) {
			logger.error("Configuration error: {} ", e.getMessage());
}
```
which should lead to error messages like ```Configuration error: Invalid network.channel (-1000 is out of range)```

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,566 @@
+	}
+
+	/**
+	 * @{inheritDoc
```
Correct tag

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,566 @@
+
+	/**
+	 * @{inheritDoc
+	 */
```
correct tag

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,566 @@
+			network.addListener(networkListener);
+			network.open();
+			logger.info("Opened network");
+//			setProperlyConfigured(true);
```
remove commented-out code

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,566 @@
+			logger.error("Error closing panStamp network: " + e.getMessage(), e);
+		} finally {
+//			setProperlyConfigured(false);
+		}
```
remove empty finally block

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,566 @@
+				setup.setChannel(cfg.networkChannel);
+			}
+			if (cfg.networkSecurityOption != -1) {
+				// FIXME
```
what should happen here?

GideonLeGrange replied:
I've removed the securityOption setting, since the underlying panStamp library does not currently support the functionality. When the library does support it, I'll bring it back.

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,566 @@
+				logger.info("Configured XML directory to {}", cfg.xmlDir);
+			} else {
+				logger.error(
+						"Devices file in XML directory {} can not be read",
```
Change "Devices" to "devices.xml" so user knows specifically the file name that cannot be read.

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,566 @@
+
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	private void addEndpoint(Endpoint ep) {
+		logger.debug("addEndpoint({}->{})", PanStampConversions.toString(ep));
```
Missing an argument for log formatting.

GideonLeGrange replied:
The debug formatting here was fixed, and I put in some effort to make it consistent in other places.

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,566 @@
+		return configs;
+	}
+
+	/* update all items associated with an endpoint with it's new value */
```
format comment in JavaDoc style.

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBinding.java:
```diff
@@ -0,0 +1,566 @@
+		}
+	};
+
+	/*
```
format comment in JavaDoc style (two asterisks).

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBindingSettings.java:
```diff
@@ -0,0 +1,143 @@
+			return !serialDiffers(other)
+					&& (networkChannel == other.networkChannel)
+					&& (networkId == other.networkId)
+					&& (networkDeviceAddress == other.networkDeviceAddress)
```
replace these 3 lines with !networkDiffers(other) to eliminate duplicate logic

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBindingSettings.java:
```diff
@@ -0,0 +1,143 @@
+					&& (networkSecurityOption == other.networkSecurityOption)
+					&& xmlDir.equals(other.xmlDir)
+					&& (debugPort == other.debugPort)
+					&& (debugEnabled == other.debugEnabled);
```
replace above 2 lines with !debugDiffers to eliminate duplicate code

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampBindingSettings.java:
```diff
@@ -0,0 +1,143 @@
+					&& (networkId == other.networkId)
+					&& (networkDeviceAddress == other.networkDeviceAddress)
+					&& (networkSecurityOption == other.networkSecurityOption)
+					&& xmlDir.equals(other.xmlDir)
```
replace with !directoriesDiffers(other) to avoid duplication

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampConversions.java:
```diff
@@ -0,0 +1,190 @@
+		throw new ValueException(String.format(
+				"Unsopported command type '%s' for string endpoint", cmd
+						.getClass().getSimpleName()));
+	}
```
Can't you use `.toString()` on any `Command`, or is there a reason to not do that in this case?

GideonLeGrange replied:
I'm not exactly sure what is right here. 

```toString(Command cmd)``` is written to complement ```toBoolean```, ```toLong```, ```toInteger``` and ```toNumber``` which are all used to convert some type (or types) of ```Command``` into a value that can be sent to a panStamp, or otherwise fail. The idea is that if the binding does not like the conversion implied by the ```Command``` and ```Endpoint``` (for example, trying to send an ```DecimalType``` to an endpoint which represents a binary output). 

I see your argument that anything could be converted to a String and passed to a String endpoint. The problem I have with this is that it may be an error on the user's side which now will not be caught and logged, while junk data is silently sent to the network. 

I don't think this is a big mistake either way though, but would prefer to leave it for now.

watou replied:
Since a `StringType` command could contain any string data including junk, then I am missing how `DecimalType.toString()` could be worse than `StringType.toString()`.  But if my suggestion could introduce junk data to the network, then definitely leave it as is.

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampConversions.java:
```diff
@@ -0,0 +1,190 @@
+	 * @return The integer value
+	 * @throws ValueException If a conversion isn't possible 
+	 */	
+	static Long toInteger(Command cmd) throws ValueException {
```
would this method not better be named `toLong`?

GideonLeGrange replied:
Yes, thanks.

watou commented on bundles/binding/org.openhab.binding.panstamp/src/main/java/org/openhab/binding/panstamp/internal/PanStampGenericBindingProvider.java:
```diff
@@ -0,0 +1,155 @@
+	@Override
+	public void processBindingConfiguration(String context, Item item,
+			String bindingConfig) throws BindingConfigParseException {
+		super.processBindingConfiguration(context, item, bindingConfig);
```
move super call to end of method, so that any preceding exception thrown will skip superclass's processing of bad bindingConfig

watou commented:
Hi @GideonLeGrange, I've made a few more minor review comments on the pull request.  Also, could you use the openHAB code formatting (Preferences->Java->Code Style->Formatter) and reformat to that (max line length at 120, for example).  There is new activity [on the forum](https://community.openhab.org/t/new-binding-panstamp-wireless-arduino-modules/2731/18?u=watou) with a user hitting a problem with serial port access; perhaps you could comment there?  Thanks very much for your excellent work!  Regards, John

GideonLeGrange commented:
Thanks for your code review. I've fixed all the errors you pointed out. I am busy testing and will update the pull request once done.

watou commented:
Thank you @GideonLeGrange for your excellent work.  Could you also add this binding to `distribution/pom.xml` please?

GideonLeGrange commented:
@watou I'm having trouble getting Eclipse to find the openHAB code formatter. I followed your instructions but there is no openHAB formatter. I downloaded and ran the yoxos installer as described in the wiki and this installed different version of Eclipse, once again with no formatter.

Any advice is appreciated. I would hate for code formatting to cause my binding to be excluded 😕

watou commented:
Hi @GideonLeGrange, I know I didn't do anything in my Eclipse to explicitly put in place the openHAB code formatter.  I used the openHAB.yoxos as described in the wiki.  [Here](https://dl.dropboxusercontent.com/u/4286376/formatter.jpg) is what I see in my Eclipse.  If it doesn't appear for you, I pressed the "Export All..." button in mine and it produced [this file](https://dl.dropboxusercontent.com/u/4286376/formatter.xml).

Also, please load up openHAB 2 and add your JAR to its addons to verify that it works there.  This is a requirement for new addons.  Looking at your code, I suspect you will have success quickly.

watou commented:
Hi @GideonLeGrange, any progress with the code formatting, user feedback or OH2 testing?  Thank you!

GideonLeGrange commented:
@watou Thank you for your help.

* I've imported your exported formatter.xml, reformatted my code and it worked. 
* I've tested using openHAB 2.0 and the binding works. 
* I've had some user feedback (as in the [forum thread](https://community.openhab.org/t/new-binding-panstamp-wireless-arduino-modules/2731/23)) and the binding seems to work for at least some others.

watou commented:
That's great, @GideonLeGrange !  If you can perform a top-level `mvn clean install` including this branch, do you get a successful build including your binding?  If so, and you only made formatting changes, then I think this PR should be ready to merge.

watou commented:
(...after being squashed into one commit and resolving the conflicts.)  Please let me know and thanks!

hakan42 commented on bundles/binding/org.openhab.binding.panstamp/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,31 @@
+Bundle-DocURL: http://www.openhab.org
+Bundle-RequiredExecutionEnvironment: J2SE-1.5
+Service-Component: OSGI-INF/binding.xml, OSGI-INF/genericbindingprovider.xml
+Bundle-ClassPath: .,
```
While not strictly required for OH1, could you add an about.html similiar to https://github.com/openhab/openhab2/blob/master/addons/binding/org.openhab.binding.avmfritz/about.html showing the additional libraries you are using and the licenses they are distributed with?

hakan42 commented:
@GideonLeGrange , are you still active on the PR?

watou commented:
I would like to finish my review and merge your binding, @GideonLeGrange, so please let us know and thank you!

watou commented:
Hello @GideonLeGrange, we are nearly out of time to consider including your binding in openHAB 1.8.  Would you be able to make final changes, tests, squash the commit into one, and verify the accuracy/completeness of the companion wiki page, in the coming days?  Many thanks for past and future efforts!

watou commented:
...also, please add the binding to binding/pom.xml and distribution/pom.xml in this PR.  Thank you!

GideonLeGrange commented:
I've got a bit of time this week and will fix the above. I'll be finished before the 1st

watou commented:
Thanks very much, @GideonLeGrange !

GideonLeGrange commented:
# Recap of this PR:

* Tested binding on openhab 1.7
* Tested binding on openhab 2.0 alpha2
* Created wiki page https://github.com/openhab/openhab/wiki/panStamp-Binding
* Fixed a long list of concerns raised by @watou
* Formatted code using openHAB code formatter
* Added to distribution/pom.xml 
* Reviewed the wiki page to make sure it is current
* Added about.html as requested by @hakan42
* Did top-level mvn clean install successfully 
* Did squash on pull request

GideonLeGrange commented:
@watou I'm done, except that I'm having issues with the squash. I tried the squash as per this document https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request but run into two problems:
* There seems to be an issue with distribution/pom.xml where I keep having to fix merge conflicts when rebasing. 
* I still see multiple commits at the top of the PR. It isn't clear to me if this is correct or indicative of a non-squashed commit.

watou commented:
Here is how I would go about rebasing and squashing.  In your local fork:
```bash
git checkout master
git fetch upstream
git merge upstream/master
# your local master is now same as openhab repo
git push origin master
# up-to-dateness is now reflected in your fork at github
git checkout panstamp-pr
git merge-base panstamp-pr master
# copy the return string to clipboard
git rebase -i <the string returned above>
# fix any merge conflicts  between >>>> and <<<< markers
# change 2nd and subsequent pick to squash
# force push your branch to github:
git push -f origin panstamp-pr
# go to github and see if the PR looks correct
```

GideonLeGrange commented:
Thank you @watou 

Squash completed as you described. As far as I'm concerned, the binding should now be ready for 1.8.0.

GideonLeGrange commented:
@watou Squash completed, I think this should be it.

hakan42 commented:
@GideonLeGrange sorry to come up with a last comment but I just noticed that you reference  (and bring the binary for it) slf4j-api-1.7.7 in your pull request. Is that necessary? Is it possible to remove that dependency and use whatever the Openhab core provides to you?

watou commented:
Good catch, @hakan42.  @GideonLeGrange, if you could remove that lib, make sure `mvn clean install` still works in your project, make sure the resulting (and now smaller) binding JAR still works correctly, commit the change, squash the commits, and `git push -f origin <your branch>`, we should be good to merge.  

Also, please promote your work, if you don't mind, at https://community.openhab.org, also describing the PanStamp system and its features and benefits.  The binding will be available after the next nightly build after the merge, and also in openHAB 1.8.  

Thanks for your perseverance!

watou commented on bundles/binding/org.openhab.binding.panstamp/about.html:
```diff
@@ -0,0 +1,49 @@
+		<strong>slf4j 1.7.7</strong> <br/><br/>
+		<a href="https://github.com/GideonLeGrange/panstamp-java">The Simple Logging Facade for Java</a> obtained from Maven central repository under 
+		<a href="http://www.slf4j.org/license.html">a copy of the MIT License</a>.
+		</em></p>
```
remove above section

watou commented on bundles/binding/org.openhab.binding.panstamp/build.properties:
```diff
@@ -0,0 +1,9 @@
+               .,\
+               OSGI-INF/,\
+               lib/nrjavaserial-3.9.3.1.jar,\
+               lib/slf4j-api-1.7.7.jar,\
```
remove JAR

watou commented on bundles/binding/org.openhab.binding.panstamp/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,31 @@
+Service-Component: OSGI-INF/binding.xml, OSGI-INF/genericbindingprovider.xml
+Bundle-ClassPath: .,
+ lib/nrjavaserial-3.9.3.1.jar,
+ lib/slf4j-api-1.7.7.jar,
```
remove above JAR

watou commented on bundles/binding/org.openhab.binding.panstamp/.classpath:
```diff
@@ -0,0 +1,11 @@
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER"/>
+	<classpathentry kind="lib" path="lib/nrjavaserial-3.9.3.1.jar"/>
+	<classpathentry kind="lib" path="lib/slf4j-api-1.7.7.jar"/>
```
remove above entry

watou commented on distribution/pom.xml:
```diff
@@ -1377,6 +1377,18 @@
+		<artifactId>org.openhab.io.caldav</artifactId>
+		<version>${project.version}</version>
+		<type>jar</type>
+	</dependency>
```
oops. ^^^  Should not be in your check-in!

GideonLeGrange replied:
That was some sort of merge problem, not sure why merging upstream/master brought that in. Wil check it out.

GideonLeGrange commented:
@hakan42 @watou Removing I've removed all external jars except for the panstamp-java library. I found gnu.io (provided by nrjavaserial) being exported by one of the transport modules, so dependencies are stripped to the absolute minimum. 

@watou I've removed that spurious reference to io.caldav too - it seems to have been bad dependency ordering in the pom on my part, probably happened when resolving merge conflicts. 

Commits squashed.

watou commented:
Nice wiki page.  I've made a small typo correction to it.  Many thanks.

hakan42 commented:
Very nice, thank you for your efforts.
